fabric-client-gametest-api-v1/src/client/java/net/fabricmc/fabric/api/client/gametest/v1/ClientGameTestContext.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.gametest.v1;
+
+import java.util.function.Predicate;
+
+import org.apache.commons.lang3.function.FailableConsumer;
+import org.apache.commons.lang3.function.FailableFunction;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.SharedConstants;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.gui.screen.Screen;
+
+/**
+ * Context for a client gametest containing various helpful functions and functions to access the game. Functions in
+ * this class can only be called on the client gametest thread.
+ */
+@ApiStatus.NonExtendable
+public interface ClientGameTestContext {
+	/**
+	 * Used to specify that a wait task should have no timeout.
+	 */
+	int NO_TIMEOUT = -1;
+
+	/**
+	 * The default timeout for wait tasks.
+	 */
+	int DEFAULT_TIMEOUT = 10 * SharedConstants.TICKS_PER_SECOND;
+
+	/**
+	 * Runs a single tick and waits for it to complete.
+	 */
+	void waitTick();
+
+	/**
+	 * Runs {@code ticks} ticks and waits for them to complete.
+	 *
+	 * @param ticks The amount of ticks to run
+	 */
+	void waitTicks(int ticks);
+
+	/**
+	 * Waits for a predicate to be true. Fails if the predicate is not satisfied after {@link #DEFAULT_TIMEOUT} ticks.
+	 *
+	 * @param predicate The predicate to check
+	 */
+	void waitFor(Predicate<MinecraftClient> predicate);
+
+	/**
+	 * Waits for a predicate to be true. Fails if the predicate is not satisfied after {@code timeout} ticks. If
+	 * {@code timeout} is {@link #NO_TIMEOUT}, there is no timeout.
+	 *
+	 * @param predicate The predicate to check
+	 * @param timeout The number of ticks before timing out
+	 */
+	void waitFor(Predicate<MinecraftClient> predicate, int timeout);
+
+	/**
+	 * Waits for the given screen class to be shown. If {@code screenClass} is {@code null}, waits for the current
+	 * screen to be {@code null}. Fails if the screen does not open after {@link #DEFAULT_TIMEOUT} ticks.
+	 *
+	 * @param screenClass The screen class to wait to open
+	 */
+	void waitForScreen(@Nullable Class<? extends Screen> screenClass);
+
+	/**
+	 * Opens a {@link Screen} on the client.
+	 *
+	 * @param screen The screen to open
+	 * @see MinecraftClient#setScreen(Screen)
+	 */
+	void setScreen(@Nullable Screen screen);
```

Comment:
Question: Would a `Suppler<Screen>` be better? Im not sure if there could be any issues constructing the screen off thread?

Reply:
Probably. I'll change it. The whole idea of "do everything on the client thread" is erring on the defensive side, the only problem it solves is when Mojang or a mod checks `Thread.currentThread()` (which incidentally happens inside `MinecraftClient.setScreen`) or uses a `ThreadLocal`.